### PR TITLE
Fix: Virtual Product Status in Sub-Orders

### PR DESCRIPTION
**Pull request by mralaminahamed**
### All Submissions:

* [x] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [x] My code satisfies feature requirements
* [x] My code is tested
* [x] My code passes the PHPCS tests
* [x] My code has proper inline documentation
* [ ] I've included related pull request(s) (optional)
* [ ] I've included developer documentation (optional)
* [x] I've added proper labels to this pull request

### Changes proposed in this Pull Request:

1. Implemented a fix to update the status of virtual products correctly within sub-orders.
2. Modified the relevant code to ensure that the status of virtual products is updated to "completed" when necessary.
3. Incorporated the necessary checks to verify if the order contains downloadable and virtual products before updating the status.

### Related Pull Request(s)

* Full PR Link

### Closes 

* Closes getdokan/dokan#2106

### How to test the changes in this Pull Request:

* Steps or issue link  getdokan/dokan#2106

### Changelog entry

* Fix: The status of virtual products in sub-orders does not update accordingly*

Detailed Description of the pull request. What was previous behaviour and what will be changed in this PR.

### Before Changes

The issue before changes with screenshots : https://drive.google.com/open?id=1cDLxu57UVhZbA-LA0Z9kKHeFEekpsEus&usp=drive_fs.

### After Changes

The issue after changes with screenshot: https://drive.google.com/open?id=1cG3Kmp-2kPLYqTdz42aew6QRfNb6mIfJ&usp=drive_fs.

### Feature Video (optional)

Link of detailed video if this PR is for a feature.

### PR Self Review Checklist:

* Code is not following code style guidelines
* Bad naming: make sure you would understand your code if you read it a few months from now.
* KISS: Keep it simple, Sweetie (not stupid!).
* DRY: Don't Repeat Yourself.
* Code that is not readable: too many nested 'if's are a bad sign.
* Performance issues
* Complicated constructions that need refactoring or comments: code should almost always be self-explanatory.
* Grammar errors.

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.

> As a contributor, your responsibility is to learn from suggestions and iterate your pull request should it be needed based on feedback. Seek to collaborate and produce the best possible contribution to the greater whole.

* [ ] Correct — Does the change do what it’s supposed to? ie: code 100% fulfilling the requirements?
* [ ] Secure — Would a nefarious party find some way to exploit this change? ie: everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities?
* [ ] Readable — Will your future self be able to understand this change months down the road?
* [ ] Elegant — Does the change fit aesthetically within the overall style and architecture?

### CODE REVIEW

**Review by shohag121 — APPROVED** [review]

**Comment by shashwatahalder01** [comment]
@mralaminahamed 
Scenario:
virtual product order status (sub-order) is completed  `(should be on hold)` when paid with direct bank payment.
N.B. order status should only be converted to completed when product is virtual or downloadable and primary order status is  processing

image

**Comment by shashwatahalder01** [comment]
Scenario:
virtual product sub-order status is processing (instead of completed) when paid with Paypal marketplace

Steps to reproduce:
1.place a multi-vendor order where one is physical and another is a virtual product

**Comment by mralaminahamed** [comment]
@shashwatahalder01 vai, i'm deeply debugging the issue and conducting research on potential solutions based on the current problem.

**Comment by mralaminahamed** [comment]
@shashwatahalder01 vai, for digital products, we update the status of sub-orders when the parent order is paid. Please double-check the last commit to ensure everything is functioning as expected.

If you encounter any issues or discrepancies, kindly let me know, and I'll be happy to investigate further.

File: includes/Order/Hooks.php [hunk 1]
```diff
@@ -263,10 +263,19 @@ public function on_sub_order_change( $order_id, $old_status, $new_status, $order
         // return if any child order is not completed
         $all_complete = true;
 
+        // Exclude manual gateways from auto complete order status update for digital products.
+        $excluded_gateways = apply_filters( 'dokan_order_status_update_exclusion_payment_gateways', array( 'bacs', 'cheque', 'cod' ) );
+
         foreach ( $sub_orders as $sub_order ) {
+            // if the order is a downloadable and virtual product, then we need to set the status to complete
+            if ( $order->is_paid() && ! in_array( $order->get_payment_method(), $excluded_gateways, true ) && ! $sub_order->needs_processing() ) {
+                $sub_order->set_status( 'completed', __( 'Mark as completed when it has digital products.', 'dokan-lite' ) );
```
Comment (mrabbani) [inline includes/Order/Hooks.php:272]: Pls change the note to **Marked as completed because it contains digital products.**

**Review by mrabbani — CHANGES_REQUESTED** [review]

File: includes/Order/Hooks.php (hunk 1 above)
Comment (mralaminahamed) [reply includes/Order/Hooks.php]: vai, order note updated.

**Comment by mrabbani** [comment]
@mralaminahamed vai, Pls check the following cases:
- [ ] Orders will not be completed for manual payments.
- [ ] All orders will be completed if all sub-orders contains digital products.
- [ ] Order will not be completed if it contains both physical and digital products.

**Comment by shashwatahalder01** [comment]
Scenario:
order note missing in one of the sub-orders if both contain virtual products.

Steps to reproduce:
1. place a multivendor order where both order contain virtual products.

Screenshot:
Screenshot at Apr 19 13-12-40
Screenshot at Apr 19 13-13-28
Screenshot at Apr 19 13-14-23
Screenshot at Apr 19 13-15-49

**Comment by shashwatahalder01** [comment]
Scenario:
Suborder has multiple same order note.

Steps to reproduce:
1. place a multivendor order where all orders contain virtual products. [3 digital product is used]

Screenshot:
Screenshot at Apr 19 15-07-07

File: includes/Order/Hooks.php [hunk 2]
```diff
@@ -263,10 +263,19 @@ public function on_sub_order_change( $order_id, $old_status, $new_status, $order
         // return if any child order is not completed
         $all_complete = true;
 
+        // Exclude manual gateways from auto complete order status update for digital products.
+        $excluded_gateways = apply_filters( 'dokan_order_status_update_exclusion_payment_gateways', array( 'bacs', 'cheque', 'cod' ) );
```
Comment (mrabbani) [inline includes/Order/Hooks.php:267]: Plse rename `dokan_order_status_update_exclusion_payment_gateways` filter name to `dokan_excluded_payment_gateways_on_order_status_update`

**Review by mrabbani — CHANGES_REQUESTED** [review]

**Comment by mralaminahamed** [comment]
@shashwatahalder01 vai, issues are fixed. please test again.

**Comment by StalinDurjo** [comment]
> Scenario: order note missing in one of the sub-orders if both contain virtual products.
> 
> Steps to reproduce:
> 
> 1. place a multivendor order where both order contain virtual products.
> 
> Screenshot: Screenshot at Apr 19 13-12-40 Screenshot at Apr 19 13-13-28 Screenshot at Apr 19 13-14-23 Screenshot at Apr 19 13-15-49

@mralaminahamed bhai, this issue is not fixed.

**Comment by mralaminahamed** [comment]
As per the decision with @mrabbani, we have removed the duplicate completion message from the order note.

**Pull request merged**